File: asys-runtime/test/fixtures.mjs
import { mkdir } from 'node:fs/promises';
import { join } from 'node:path';
import { Queue } from '../javascript/queue.mjs';

export class PreparedQueue extends Queue {
  async resubmit(previous, id) {
    const request = await this.request(previous);
    return this.submit(request.type, id, { input: request.input, args: request.args,
      metadata: { ...request.metadata, retry_of: previous }, workspace: (await this.paths(previous)).workspace });
  }
  executionDirectory(id) { return join(this.root, 'executions', id, 'job'); }
  workspace(id) { return join(this.root, 'executions', id, 'workspace'); }
  async submit(type, id, options = {}) {
    const directory = this.executionDirectory(id), workspace = this.workspace(id);
    await mkdir(directory, { recursive: true });
    await mkdir(workspace, { recursive: true });
    return super.submit(type, id, { directory, workspace, ...options });
  }
}

File: asys-runtime/test/fixtures.py
from pathlib import Path
from asys_runtime import Queue


class PreparedQueue(Queue):
    """The test caller prepares execution storage before submitting work."""
    def submit(self, job_type, job_id, **options):
        storage = self.root / "executions" / job_id
        job, workspace = storage / "job", storage / "workspace"
        job.mkdir(parents=True, exist_ok=True)
        workspace.mkdir(parents=True, exist_ok=True)
        return super().submit(job_type, job_id, **{"directory": job, "workspace": workspace, **options})
